Refactor ConfigReloadWatcher component (#1363)
kyma-project/kyma-infrastructure-manager: docs/operator/config-reload-watcher.md

# ConfigReloadWatcher

## Overview

> The `ConfigReloadWatcher` is a controller that monitors changes to Kubernetes configuration resources (ConfigMaps, Secrets, and ClusterTrustBundles) on the Kyma Control Plane (KCP).When a watched resource is modified, the controller triggers re-reconciliation of affected Runtime CRs by annotating them with a force-reconcile annotation.

Reviewer: ```suggestion
The `ConfigReloadWatcher` is a controller that monitors changes to Kubernetes configuration resources (ConfigMaps, Secrets, and ClusterTrustBundles) on Kyma Control Plane (KCP). When a watched resource is modified, the controller triggers re-reconciliation of affected Runtime custom resources (CRs) by annotating them with a force-reconcile annotation.
```

> This mechanism ensures that changes to shared configuration, such as API server ACL rules or runtime bootstrapper settings, are propagated to all relevant Kyma runtimes without requiring manual intervention.

Reviewer: ```suggestion
This mechanism ensures that changes to shared configuration, such as API server Access Control List (ACL) rules or Runtime Bootstrapper settings, are propagated to all relevant Kyma runtimes without requiring manual intervention.
```

Reviewer: Hope I understand the ACL and the reference to [Runtime Bootstrapper](https://github.com/kyma-project/rt-bootstrapper/blob/main/docs/contributor/README.md) correctly.

Reviewer 2: Yes :D

## Activation

The controller is registered only when at least one of the following feature flags is enabled:

> - `--api-server-acl-enabled` - enables the shoot API server ACL extender

Reviewer: ```suggestion
- `--api-server-acl-enabled` - enables the Shoot API server ACL extender
```

> - `--runtime-bootstrapper-enabled` - enables the runtime bootstrapper

Reviewer: ```suggestion
- `--runtime-bootstrapper-enabled` - enables Runtime Bootstrapper
```

If neither flag is set, the controller is not created and no watches are registered.

## Watched Resources

> Depending on which features are enabled, the controller watches different sets of resources:

Reviewer: ```suggestion
The resources watched depend on the enabled feature.
```

> ### When `--runtime-bootstrapper-enabled` is set

Reviewer: ```suggestion
- For `--runtime-bootstrapper-enabled`, the controller watches the following resources:
```

Reviewer: Please consider using bullet points instead of sections. If you do, please indent the sections' content so it is clearly visible as belonging to the bullet points.

| Resource Type | Name Source | Namespace |
|:---|:---|:---|
| ConfigMap | `--runtime-bootstrapper-kcp-config-name` | `kcp-system` |
| ConfigMap | `--runtime-bootstrapper-manifests-config-map-name` | `kcp-system` |
| ClusterTrustBundle | `--runtime-bootstrapper-kcp-cluster-trust-bundle` (optional) | cluster-scoped |
| Secret | `--runtime-bootstrapper-kcp-pull-secret-name` (optional) | `kcp-system` |

The ClusterTrustBundle and pull Secret watches are only registered when the corresponding flag value is non-empty.

> ### When `--api-server-acl-enabled` is set

Reviewer: ```suggestion
- For `--api-server-acl-enabled`, the controller watches the following resource:
```

| Resource Type | Name Source | Namespace |
|:---|:---|:---|
| ConfigMap | `converterConfig.kubernetes.kubeApiServer.acl.configMapName` | `kcp-system` |

## Reconciliation Flow

When a watched resource is updated, the following steps occur:

1. A watched resource (ConfigMap, Secret, or ClusterTrustBundle) is updated.

> 2. The `ObjectUpdatedPredicate` filters the event - only Update events with a matching name and namespace pass through.

Reviewer: ```suggestion
2. The `ObjectUpdatedPredicate` filters the event - only **Update** events with a matching name and namespace pass through.
```

3. The `ConfigReloadWatcher.Reconcile` method is called. It lists all Runtime CRs in the `kcp-system` namespace.

> 4. For each Runtime, two checks are applied:

Reviewer: 1. It's not quite clear to me what the two checks are. One is RuntimePredicate, right? And what is the other one?
2. You use "Runtime" quite frequently in this section. What is it? Is it "Runtime CR" or a "Kyma runtime instance"?

Reviewer 2: I made changes to that section. I hope it is clearer now.

>    - The `RuntimePredicate` determines if this Runtime should react to the specific config change ([see below](#runtime-predicate)).

Reviewer: ```suggestion
   - The `RuntimePredicate` determines if this Runtime should react to the specific config change (see [Runtime Predicate](#runtime-predicate)).
```

   - If the Runtime already has the `force-patch-reconciliation` annotation, it is skipped.

> 5. Eligible Runtimes are patched via server-side apply with the annotation `operator.kyma-project.io/force-patch-reconciliation=true`.

Reviewer: ```suggestion
5. Eligible Runtimes are patched through server-side apply with the annotation `operator.kyma-project.io/force-patch-reconciliation=true`.
```

> 6. The Runtime Controller detects the annotated CR and triggers a full re-reconciliation.

Reviewer: What is Runtime Controller?

Reviewer 2: It is the controller that manages the RuntimeCR

## Runtime Predicate

Not all Runtimes need to react to every configuration change. The `RuntimePredicate` function determines whether a specific Runtime should be re-reconciled for a given configuration change:

> - If the triggering resource is the ACL ConfigMap, the predicate calls `AclNeedsToBeEnabled`, which returns `true` only when:

Reviewer: ```suggestion
- If the triggering resource is the ACL ConfigMap, the predicate calls `AclNeedsToBeEnabled`, which returns `true` only when the following conditions are met:
```

Reviewer: It would be good to say if all or only one of the conditions must be met.

  - `--api-server-acl-enabled` is `true`
  - The Runtime's provider type is AWS or Azure
  - The Runtime has a non-empty ACL AllowedCIDRs list

> - For all other resources (runtime bootstrapper ConfigMaps, Secrets, ClusterTrustBundles), the predicate returns `true` - all Runtimes are re-reconciled.

Reviewer: ```suggestion
- For all other resources (Runtime Bootstrapper ConfigMaps, Secrets, ClusterTrustBundles), the predicate returns `true` - all Runtimes are re-reconciled.
```

## Event Filtering

> The controller uses `ObjectUpdatedPredicate` to filter Kubernetes watch events. Only **Update** events are processed Create, Delete, and Generic events are ignored.

Reviewer: ```suggestion
The controller uses `ObjectUpdatedPredicate` to filter Kubernetes watch events. Only **Update** events are processed. The **Create**, ** Delete **, and ** Generic ** events are ignored.
```
